bin/run.sh

```diff
@@ -15,73 +15,284 @@
+    local test_file test_name test_body description task_id trimmed
+
+    for test_file in ./*_test.odin; do
+        while IFS= read -r line; do
+            if ! "$in_test"; then
+                if [[ $line == '@(test)' ]]; then
+                    test_body=()
+                    description=''
+                    task_id=''
+                    in_test=true
+                fi
+            else
+                if [[ $line =~ ^'/// description = '(.+) ]]; then
+                    description=${BASH_REMATCH[1]}
+                elif [[ $line =~ ^'/// task_id = '(.+) ]]; then
+                    task_id=${BASH_REMATCH[1]}
+                elif [[ $line =~ ([[:alnum:]_]+)' :: proc' ]]; then
+                    test_name=${BASH_REMATCH[1]}
+                    if [[ -n $description ]]; then
+                        descriptions["$test_name"]=$description
+                    else
+                        descriptions["$test_name"]=$test_name
+                    fi
+                    [[ -n $task_id ]] && task_ids["$test_name"]=$task_id
+                    test_names+=( "$test_name" )
+                    in_proc=true
+                elif [[ $line == '}' ]]; then
+                    in_test=false
+                    in_proc=false
+                    # shellcheck disable=SC2034
+                    test_code["$test_name"]=$( join $'\n' "${test_body[@]}" )
+                elif $in_proc; then
+                    read -r trimmed <<< "$line"
+                    [[ -n $trimmed ]] && test_body+=( "$trimmed" )
+                fi
+            fi
+        done < "${test_file}"
+    done
+}
+
+# Populates the global `failure_messages` map
+parse_odin_test_output() {
+    local test_output=$1
+    local -a failure_message
+    local seen_finished=false
+    test_name=''
+
+    while IFS= read -r line; do
+        if ! $seen_finished; then
+            if [[ $line == "Finished "*" tests. "*" test"*"failed." ]]; then
+                seen_finished=true
+            fi
+        else
+            if [[ $line =~ ^" - ${package_name}."([[:alnum:]_]+)[[:blank:]]+(.*) ]]; then
+                # Start of test report
+                if [[ -n $test_name && -n "$failure_message" ]]; then
+                    failure_messages["$test_name"]=$(join $'\n' "${failure_message[@]}")
+                fi
+                test_name=${BASH_REMATCH[1]}
+                failure_message=( "${BASH_REMATCH[2]}" )
+            else
+                # Midst of a test report
+                failure_message+=( "$line" )
+            fi
+        fi
+    done <<< "$test_output"
+    if [[ -n $test_name && -n "${failure_message[*]}" && ! -v "failure_messages[$test_name]" ]]; then
+        failure_messages["$test_name"]=$(join $'\n' "${failure_message[@]}")
+    fi
+}
+
+#------------------------------------------------------------
+main() {
+    # If any required arguments is missing, print the usage and exit
+    [[ -n "$1" && -n "$2" && -n "$3" ]] || usage
```

Comment:
```suggestion
    (( $# == 3 )) || usage
```

Reply:
Just testing the number of args allows empty arguments.